Issue #81 No failure during duplicate and write error into event

TomoTsuyuki opened the pull request:
Implemented to avoid error in the middle of duplication. #81 
Event "massaction_duplicated" has been introduced for the logging.

![image](https://github.com/Syxton/moodle-block_massaction/assets/54526084/1577bafa-5e83-4bf3-8025-3cd6d3d86ecc)

dmitriim commented on classes/event/massaction_duplicated.php:
```diff
@@ -0,0 +1,71 @@
+ * @copyright   2023 Catalyst IT
+ * @license     https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class massaction_duplicated extends base {
```
We already have this event under massaction block. We don't really need to duplicate it in event name. How about renaming to **activities_duplicated** or **course_modules_duplicated**?

TomoTsuyuki replied:
I renamed it to "course_modules_duplicated".

dmitriim commented on classes/event/massaction_duplicated.php:
```diff
@@ -0,0 +1,71 @@
+        foreach ($this->other['cms'] as $srccm => $dstcm) {
+            $cms[] = 'cmid from \'' . $srccm . '\' to \'' . $dstcm . '\'';
+        }
+        $errormsg = !empty($this->other['errors']) ? " with error '" . implode("','", $this->other['errors']) : "'";
```
As said before I think better just to log a number of errors rather than each individual error. And have a separate event (e.g. **activity_duplication_failed** ) for any error message. But it's up to a maintainer to decide.

TomoTsuyuki replied:
The error details go to a new "failed" event and the "duplicated" event has just number of failed and cmid.

dmitriim commented on classes/event/massaction_duplicated.php:
```diff
@@ -0,0 +1,71 @@
+    public function get_description(): string {
+        $cms = [];
+        foreach ($this->other['cms'] as $srccm => $dstcm) {
+            $cms[] = 'cmid from \'' . $srccm . '\' to \'' . $dstcm . '\'';
```
would be also useful to see which failed and which succeeded .

TomoTsuyuki replied:
I added total of complete/fail to the description.

dmitriim commented on classes/event/massaction_duplicated.php:
```diff
@@ -0,0 +1,71 @@
+    protected function validate_data() {
+        parent::validate_data();
+
+        if (!isset($this->other['cms']) || !is_array($this->other['cms'])) {
```
should we validate $this->other['errors'] also?

TomoTsuyuki replied:
I added the new validation.

dmitriim commented:
Thanks @TomoTsuyuki  for working on it. The code looks good in general. I have seen people copping hundreds of activities. I'm a bit worried how that all will look at the logs screen. I have left my comments. 

Also the patch covers just duplicate, what about  duplicate_to_course?

TomoTsuyuki commented on classes/actions.php:
```diff
@@ -147,8 +147,16 @@ public static function duplicate(array $modules, $sectionnumber = false): void {
+            try {
+                $duplicatedmod = duplicate_module($modinfo->get_course(), $modinfo->get_cm($cmid));
+            } catch (\Exception $e) {
+                $errors[] = 'cmid:' . $cmid . '(' . $e->getMessage() . ')';
```
I push the error detail to a new "failed" event, and make simple for a "duplicate" event.

TomoTsuyuki commented:
> Thanks @TomoTsuyuki for working on it. The code looks good in general. I have seen people copping hundreds of activities. I'm a bit worried how that all will look at the logs screen. I have left my comments.
> 
> Also the patch covers just duplicate, what about duplicate_to_course?

Thanks for your review @dmitriim 
I updated the events and support duplicate_to_course.

Please let me know if they are ok.

dmitriim commented on classes/actions.php:
```diff
@@ -147,8 +147,16 @@ public static function duplicate(array $modules, $sectionnumber = false): void {
+            try {
+                $duplicatedmod = duplicate_module($modinfo->get_course(), $modinfo->get_cm($cmid));
+            } catch (\Exception $e) {
+                $errors[$cmid] = 'cmid:' . $cmid . '(' . $e->getMessage() . ')';
```
any particular reason why you don't fir an event here, but instead later looping through the errors?

TomoTsuyuki replied:
I moved the event to catch block.

dmitriim commented on classes/actions.php:
```diff
@@ -276,6 +312,26 @@ public static function duplicate_to_course(array $modules, int $targetcourseid,
             }
         }
+        $event = \block_massaction\event\course_modules_duplicated::create([
+            'context' => \context_course::instance($targetcourseid),
```
Why do you think we should trigger all those events in a context of a target course and not a source course?

TomoTsuyuki replied:
I changed to `$sourcecourseid` instead of `$targecourseid`

Syxton commented on classes/actions.php:
```diff
@@ -262,8 +286,24 @@ public static function duplicate_to_course(array $modules, int $targetcourseid,
         foreach ($idsincourseorder as $cmid) {
-            $duplicatedmod = massactionutils::duplicate_cm_to_course($targetmodinfo->get_course(), $sourcemodinfo->get_cm($cmid));
+            try {
+                $duplicatedmod = massactionutils::duplicate_cm_to_course($targetmodinfo->get_course(), $sourcemodinfo->get_cm($cmid));
```
Line exceeds 132 characters;

TomoTsuyuki replied:
Thanks for your review.
I updated the code to add a break line.

Syxton commented:
@TomoTsuyuki I'm sorry for all the delays on these proposed fixes.  I had to make some decisions on the future direction of the plugin.  We have made a fairly major change moving forward in 4.2+ so these changes and the ones in the other PR's will need rebased to the latest master commits.  Also, when making these requests, please do not make changes to the version.php and I will alter the version number once we have integrated changes.  That way we don't have constant conflicts on version numbers.  I understand that for testing the version number might need tweaked to add new settings, lang strings, etc.  Again sorry to send you back to development on all this work.  Your time invested is really valued.  Thank you.

TomoTsuyuki commented:
Hi @Syxton ,

Thanks for your comment.
I rebased my branch with master, and removed the version.php from the commit.
Are you accepting PRs for the new `MOODLE_401_STABLE` branch?
If you will maintain only master branch, we need to make our fork to update the `MOODLE_401_STABLE` branch as our clients are using 4.1 LTS.

Syxton commented:
@TomoTsuyuki The 4.1 branch still works on 4.2 and 4.3, so I will of course still accept PR's that improve and extend the lifespan of that version of the plugin.  Just create another PR for the sub 4.2 changes and submit against MOODLE_401_STABLE

TomoTsuyuki commented:
Thanks @Syxton , that's good news.
I made PRs for 401 branches.
Please review and let me know if I need to do extra work.

Syxton commented on lang/en/block_massaction.php:
A few minor things:
When is the event:massaction_duplicated string used?
Can we change the failed text from "Course modules duplicated failed" to "Course modules failed to duplicate"

Syxton replied:
@TomoTsuyuki just wanted to make sure you got notified on this.

TomoTsuyuki replied:
Thanks for your review, @Syxton !

I deleted `event:massaction_duplicated` and update `event:course_modules_duplicated_failed` string.

TomoTsuyuki replied:
I also rebased by the latest master/MOODLE_401_STABLE branch.
